core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala

```diff
@@ -540,6 +540,43 @@ class DAGSchedulerSuite extends SparkFunSuite with LocalSparkContext with TimeLi
     assert(mapStatus2(2).location.host === "hostB")
   }
 
+  test("[SPARK-32003] All shuffle files for executor should be cleaned up on fetch failure") {
+    // reset the test context with the right shuffle service config
+    afterEach()
+    val conf = new SparkConf()
+    conf.set(config.SHUFFLE_SERVICE_ENABLED.key, "true")
+    init(conf)
+
+    val shuffleMapRdd = new MyRDD(sc, 3, Nil)
+    val shuffleDep = new ShuffleDependency(shuffleMapRdd, new HashPartitioner(3))
+    val shuffleId = shuffleDep.shuffleId
+    val reduceRdd = new MyRDD(sc, 3, List(shuffleDep), tracker = mapOutputTracker)
+
+    submit(reduceRdd, Array(0, 1, 2))
+    // Map stage completes successfully,
+    // two tasks are run on an executor on hostA and one on an executor on hostB
+    completeShuffleMapStageSuccessfully(0, 0, 3, Seq("hostA", "hostA", "hostB"))
+    // Now the executor on hostA is lost
+    runEvent(ExecutorLost("hostA-exec", ExecutorExited(-100, false, "Container marked as failed")))
+
+    // The MapOutputTracker has all the shuffle files
+    val initialMapStatuses = mapOutputTracker.shuffleStatuses(shuffleId).mapStatuses
+    assert(initialMapStatuses.count(_ != null) === 3)
+    assert(initialMapStatuses.count(s => s != null && s.location.executorId == "hostA-exec") === 2)
+    assert(initialMapStatuses.count(s => s != null && s.location.executorId == "hostB-exec") === 1)
+
+    // Now a fetch failure from the lost executor occurs
+    complete(taskSets(1), Seq(
+      (FetchFailed(makeBlockManagerId("hostA"), shuffleId, 0L, 0, 0, "ignored"), null)
+    ))
+
+    // Shuffle files for hostA-exec should be lost
+    val mapStatuses = mapOutputTracker.shuffleStatuses(shuffleId).mapStatuses
+    assert(mapStatuses.count(_ != null) === 1)
+    assert(initialMapStatuses.count(s => s != null && s.location.executorId == "hostA-exec") === 0)
+    assert(initialMapStatuses.count(s => s != null && s.location.executorId == "hostB-exec") === 1)
```

Comment:
As `mapStatuses` is reflecting the underlying change it is not needed to get it in the second time:

```suggestion
    val mapStatuses = mapOutputTracker.shuffleStatuses(shuffleId).mapStatuses
    assert(mapStatuses.count(_ != null) === 3)
    assert(mapStatuses.count(s => s != null && s.location.executorId == "hostA-exec") === 2)
    assert(mapStatuses.count(s => s != null && s.location.executorId == "hostB-exec") === 1)

    // Now a fetch failure from the lost executor occurs
    complete(taskSets(1), Seq(
      (FetchFailed(makeBlockManagerId("hostA"), shuffleId, 0L, 0, 0, "ignored"), null)
    ))

    // Shuffle files for hostA-exec should be lost
    assert(mapStatuses.count(_ != null) === 1)
    assert(mapStatuses.count(s => s != null && s.location.executorId == "hostA-exec") === 0)
    assert(mapStatuses.count(s => s != null && s.location.executorId == "hostB-exec") === 1)

```

Reply:
You are absolutely right; `shuffleStatuses` in the MapOutputTracker is a mutable `Map` and `mapStatuses` in the values are mutable `Array`s. So everything is updated in place. (And I accidentally just used `initialMapStatuses` for the verification anyway! (due to careless copy-and-paste))
The reason I'd gotten `mapStatuses` twice is because I followed the pattern in the test preceding this one.
Thanks for catching this.